plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/consumedservices/orchestrator/rulebackingstore/SqlRuleStore.java

```diff
@@ -0,0 +1,119 @@
+    private final String SQL_SELECT_RULES = "select id from pde_rule where plant_description_id=?;";
+    private final String SQL_INSERT_RULE = "INSERT INTO pde_rule(id, plant_description_id) VALUES(?, ?);";
+    private final String SQL_DELETE_RULES = "DELETE FROM pde_rule where plant_description_id=?;";
+    private final String ID = "id";
+
+    private Connection connection;
+
+    /**
+     * Throws an {@code IllegalStateException} if this instance has not been'
+     * initialized.
+     */
+    private void ensureInitialized() {
+        if (connection == null) {
+            throw new IllegalStateException("SqlRuleStore has not been initialized.");
+        }
+    }
+
+    /**
+     * Initializes the rule store for use by connecting to the database and
+     * creating the necessary tables.
+     *
+     * @param driverClassName The driver class for the mysql database.
+     * @param connectionUrl   URL of the database connection.
+     * @param username        Username to use when connecting to the database.
+     * @param password        Password to use when connecting to the database.
+     */
+    public void init(
+        final String driverClassName,
+        final String connectionUrl,
+        final String username,
+        final String password
+    ) throws RuleStoreException {
+
+        Objects.requireNonNull(driverClassName, "Expected database driver name.");
+        Objects.requireNonNull(connectionUrl, "Expected connection URL.");
+        Objects.requireNonNull(username, "Expected username.");
+        Objects.requireNonNull(password, "Expected password.");
+
+        try {
+            Class.forName(driverClassName);
+            connection = DriverManager.getConnection(connectionUrl, username, password);
+            final Statement statement = connection.createStatement();
+            statement.execute(SQL_CREATE_TABLE);
+
+        } catch (final ClassNotFoundException | SQLException e) {
+            throw new RuleStoreException("Failed to initialize rule store", e);
+        }
+    }
+
+    @Override
+    public Set<Integer> readRules(final int plantDescriptionId) throws RuleStoreException {
+        ensureInitialized();
+
+        try {
+            final Set<Integer> result = new HashSet<>();
+            final PreparedStatement statement = connection.prepareStatement(SQL_SELECT_RULES);
+            statement.setInt(1, plantDescriptionId);
+            final ResultSet resultSet = statement.executeQuery();
+
+            while (resultSet.next()) {
+                result.add(resultSet.getInt(ID));
+            }
+
+            return result;
+
+        } catch (final SQLException e) {
+            throw new RuleStoreException("Failed to read rules", e);
+        }
+    }
+
+    @Override
+    public void writeRules(final int plantDescriptionId, final Set<Integer> rules) throws RuleStoreException {
```

Comment:
No transaction handling. If, for example, the 5th record writing is failed, the first 4 remains in the db. You should turn off automatic commit before the loop, commit after the loop, rollback in the catch block and turn on automatic commit in a finally block.

See https://docs.oracle.com/javase/tutorial/jdbc/basics/transactions.html for example.

Reply:
Fixed.